Bureaucracy module. Iteration 2. Hiring flow migration.

shamil-gadelshin opened the pull request:
### Features
- this PR contains migrated "hiring flow" extrinsics from the content working group module
- tests for migrated extrinsics
- fixed tests for the forum module
- the 'curator' entity was renamed to the 'worker'

### Extrinsics
+ add_worker_opening
+ accept_worker_applications
+ begin_worker_applicant_review
+ fill_worker_opening
+ withdraw_worker_application
+ terminate_worker_application
+ apply_on_worker_opening

#### Open question
**How should we create a mint for the bureaucracy instance?** 

In the content working group the mint is created during the genesis config initialization:
```
add_extra_genesis {
        config(mint_capacity): minting::BalanceOf<T>;
        // config(mint_adjustment): minting::Adjustment<BalanceOf<T>, T::BlockNumber> (add serialize/deserialize derivation for type)
        build(|config: &GenesisConfig<T>| {
            // create mint
            let mint_id = <minting::Module<T>>::add_mint(config.mint_capacity, None).expect("Failed to create a mint for the content working group");
            Mint::<T>::put(mint_id);
        });
    }
```

**Should we add a) the same config initialization or b) create a migration or c) an extrinsic?**

shamil-gadelshin committed:
Restore add_opening() method in the bureacracy module

shamil-gadelshin committed:
Add tests mock for the bureacracy module

shamil-gadelshin committed:
Fix merge

shamil-gadelshin committed:
Add hiring extrinsics for the bureacracy module

shamil-gadelshin committed:
Fix clippy warnigs

shamil-gadelshin committed:
Fix forum tests

shamil-gadelshin committed:
Move set_lead test from forum to the bureaucracy

shamil-gadelshin committed:
Add support for bureacracy module event tests

shamil-gadelshin committed:
Add error wrappers

- add error wrappers
- restore hiring module error checks

shamil-gadelshin committed:
Restore events for the hiring flow extrinsics

-restore events for the hiring flow extrinsics in the bureaucracy module

shamil-gadelshin committed:
Restore most checks for hiring flow extrinsics

- restore most checks for hiring flow extrinsics in the bureaucracy module

shamil-gadelshin committed:
Add tests for add_curator_opening() in the bureaucracy

shamil-gadelshin committed:
Add tests for bureaucracy::accept_curator_applications()

shamil-gadelshin committed:
Add test for bureaucracy::apply_on_curator_opening

shamil-gadelshin committed:
Simplify events fixture for the bureaucracy module

shamil-gadelshin committed:
Update bureaucracy::add_curator_opening() test

shamil-gadelshin committed:
Update bureaucracy::apply_on_curator_opening test

- add content check
- add tests for invalid conditions

shamil-gadelshin committed:
Add tests for bureaucracy::apply_on_curator_opening()

shamil-gadelshin committed:
Move error values in the bureuacracy module

shamil-gadelshin committed:
Rename wrapper_error.rs

- renare wrapper_error.rs to errors.rs in the bureaucracy module

shamil-gadelshin committed:
Add withdraw_curator_application() with tests

shamil-gadelshin committed:
Add terminate_curator_application() with tests

shamil-gadelshin committed:
Add begin_curator_applicant_review() extrinsic and tests

shamil-gadelshin committed:
Add fill_curator_opening() sketch

shamil-gadelshin committed:
Add some tests for fill_curator_opening()

shamil-gadelshin committed:
Rename the ‘curator’ entity to the ‘worker’

shamil-gadelshin committed:
Restore commented code in fill_worker_opening()

shamil-gadelshin committed:
Add tests for the bureaucracy::fill_worker_opening()

bedeho commented on runtime-modules/bureaucracy/src/constraints.rs:
```diff
@@ -0,0 +1,40 @@
+use codec::{Decode, Encode};
```
Can we instead use a version of this from the common module that you added?

shamil-gadelshin replied:
Sure!

bedeho reviewed (changes requested):
Pretty much perfect, nice job!

bedeho commented:
For the mint, let's wait until we get to the issue of when we want to deploy this.

shamil-gadelshin committed:
Move InputValidationLengthConstraint to the common module

- move InputValidationLengthConstraint to the common module from the forum
- update forum
- update content working group

shamil-gadelshin committed:
Refactor InputValidationLengthConstraint in the bureaucracy

- delete constraints module
- use common::constraints::InputValidationLengthConstraint instead

shamil-gadelshin commented:
I refactored InputValidationLengthConstraint:
- moved out from the forum to the common module
- updated forum, content_working_group, bureaucracy modules
